public/app/plugins/datasource/cloudwatch/components/LogsQueryField.tsx

```diff
@@ -0,0 +1,321 @@
+// Libraries
+import React, { ReactNode } from 'react';
+import intersection from 'lodash/intersection';
+
+import {
+  QueryField,
+  SlatePrism,
+  FormField,
+  TypeaheadInput,
+  TypeaheadOutput,
+  BracesPlugin,
+  Select,
+  MultiSelect,
+} from '@grafana/ui';
+
+// Utils & Services
+// dom also includes Element polyfills
+import { Plugin, Node } from 'slate';
+import syntax from '../syntax';
+
+// Types
+import { ExploreQueryFieldProps, AbsoluteTimeRange, SelectableValue } from '@grafana/data';
+import { CloudWatchQuery, CloudWatchLogsQuery } from '../types';
+import { CloudWatchDatasource } from '../datasource';
+import Prism, { Grammar } from 'prismjs';
+import { CloudWatchLanguageProvider } from '../language_provider';
+import { css } from 'emotion';
+
+export interface CloudWatchLogsQueryFieldProps extends ExploreQueryFieldProps<CloudWatchDatasource, CloudWatchQuery> {
+  absoluteRange: AbsoluteTimeRange;
+  onLabelsRefresh?: () => void;
+  ExtraFieldElement?: ReactNode;
+  syntaxLoaded: boolean;
+  syntax: Grammar;
+}
+
+const containerClass = css`
+  flex-grow: 1;
+  min-height: 35px;
+`;
+
+const rowGap = css`
+  gap: 3px;
+`;
+
+interface State {
+  selectedLogGroups: Array<SelectableValue<string>>;
+  availableLogGroups: Array<SelectableValue<string>>;
+  loadingLogGroups: boolean;
+  regions: Array<SelectableValue<string>>;
+  selectedRegion: SelectableValue<string>;
+}
+
+export class CloudWatchLogsQueryField extends React.PureComponent<CloudWatchLogsQueryFieldProps, State> {
+  state: State = {
+    selectedLogGroups:
+      (this.props.query as CloudWatchLogsQuery).logGroupNames?.map(logGroup => ({
```

Comment:
Should the props here be `ExploreQueryFieldProps<CloudWatchDatasource, CloudWatchLogsQuery>` then so this is recognised automatically?

Reply:
I would prefer this but the typing of `ExploreQueryFieldProps`, `QueryEditorProps`, and `DataSourceAPI` are such that the `query` prop always needs to extend the `TQuery` type parameter used when defining a particular datasource, even if you know a component like the log query field is actually only a subtype of `TQuery`.
I tried changing the typing to be more flexible in this regard but didn't have much luck